### ISSUE TITLE
libstableR hessian hold alpha constant bonks for alpha==2 (PPN)

### ISSUE THREAD

**Issue by swihart** [swihart/gnlrim#12]
```
##############################################################
##############################################################
##############################################################
## this section tests the libstableR implementation of
## subgauss and compares it to stabledist implementation of
## subgauss and compares to PPN with [dp]norm
##############################################################
##############################################################
##############################################################

dose <- c(c(9,12,4,9,11,10,2,11,12,9,9,9,4,9,11,9,14,7,9,8), 3+c(9,12,4,9,11,10,2,11,12,9,9,9,4,9,11,9,14,7,9,8))
y_01 <- rep(as.numeric(c(8.674419, 11.506066, 11.386742, 27.414532, 12.135699,  4.359469,
                         1.900681, 17.425948,  4.503345,  2.691792,  5.731100, 10.534971,
                         11.220260,  6.968932,  4.094357, 16.393806, 14.656584,  8.786133,
                         20.972267, 17.178012) > 4),2)
id <- rep(1:8, each=5)

y_cbind = cbind(y_01, 1-y_01)

fit_PPN_scl <- gnlrim(y=y_cbind,
                      mu=~pnorm(a+b*dose+rand),
                      pmu=c(a=0,b=0),
                      pmix=c(scl=1),
                      p_uppb = c(Inf ,  Inf, Inf),
                      p_lowb = c(-Inf, -Inf,   0+1e-5),
                      distribution="binomial",
                      nest=id,
                      random="rand",
                      mixture="normal-var")
fit_PPN_scl$coefficients
fit_PPN_scl$se



#Now try reproducing PPN with **libstableR** subgauss:

fit_PPN_libstableR_scl <- gnlrim(y=y_cbind,
                                 mu=~ stable_cdf(a+b*dose+rand, c(alpha, 0, 1/sqrt(2), 0)),
                                 pmu=c(a=0,b=0, alpha=2),
                                 pmix=c(alpha=2, scl=1),
                                 p_uppb = c(Inf ,  Inf, 2, Inf),
                                 p_lowb = c(-Inf, -Inf, 2,   0+1e-5),
                                 distribution="binomial",
                                 nest=id,
                                 random="rand",
                                 mixture="libstableR-subgauss-scl",
                                 ooo=TRUE,
                                 trace=1
                                 )

## should be 0s, not NAs.
attr(fit_PPN_libstableR_scl, "details")[1,]$nhatend
bonk <- attr(fit_PPN_libstableR_scl, "details")[1,]$nhatend
bonk_no_na <- matrix(bonk[!is.na(bonk)], nrow=3)
bonk_no_na_cov <- solve(bonk_no_na)
bonk_se <- sqrt(diag(bonk_no_na_cov))
## compared these two:
bonk_se
fit_PPN_scl$se
```

```
> attr(fit_PPN_libstableR_scl, "details")[1,]$nhatend
           [,1]       [,2] [,3]       [,4]
[1,]  0.9104216   7.984049   NA  -1.122670
[2,]  7.9840491 102.693866   NA -13.913742
[3,]         NA         NA   NA         NA
[4,] -1.1226702 -13.913742   NA   2.350043
> bonk <- attr(fit_PPN_libstableR_scl, "details")[1,]$nhatend
> bonk_no_na <- matrix(bonk[!is.na(bonk)], nrow=3)
> bonk_no_na_cov <- solve(bonk_no_na)
> bonk_se <- sqrt(diag(bonk_no_na_cov))
> bonk_se
[1] 1.8696026 0.2537024 1.4758298
> fit_PPN_scl$se
[1] 1.869593 0.253700 9.009840
```

**Potential fix:  within gnlrim, identify fixed parameters as pmu=p_uppb=p_lowb and then do the subsetting of hessian before inversion based on that as opposed to based on 0 rows/cols.**

  * it is reassuring to see that the first two elements of `bonk_se` and `fit_PPN_scl$se` are the same.  the 3rd is just due to one estimating the variance and the other estimating the scale.

**Comment by swihart** [comment]
Whoa!  I just made a branch `libstableR-wrappers` and made `cdf` and `pdf` that internally forced alpha, beta, gamma to stay in their bounds and this allowed the hessian to get computed and not throw all those crazy errors.

I may want to export these so I can use them after loading package. Need to give it a little more thought...

**Comment by swihart** [comment]
So here's what I'm going to do:

  1.  Make .R files that can be roxygenized with 4 functions:
  * stable_cdf2
  * stable_pdf2
  * pstable2
  * dstable2

  2.  Then change gnlrim to use them
  3.  Then change scratch_testing.R to use them
  4.  D, B, E them.
  5.  Then commit to BRANCH and push BRANCH to github (do NOT merge to master yet)
  6. install BRANCH to cluster
   *
  7. see if I can run examples on cluster

**Comment by swihart** [comment]
Examples are running on cluster.

Some points of observation:

  *  for `method="nlminb"`, the holding a parameter constant seems to be kosher -- what I mean is that we get SEs and estimates that make sense...although this needs to be double checked.  I took a peak at gnlrim-small-batch on cluster and it seems like for really low alphas (.40 - .50) things were all over the map
  *  The codes I'm getting from gnlrim are not nlminb or nlm -- they are optimx.  optimx gives convergence code 0 if successful and 1 if iteration limit was reached.  So I just did PR on this branch with compute_hessian and compute_kkt which allows me to optimize and stop before computing hessian.  This speeds up things considerably.  Once this latest batch job is done I want to reinstall on cluster and set up new directory of ooo=TRUE and crank the iteration limit up because I notice a few 1's when peeking --

Something like:

```
fit_PPN_libstableR_scl <- gnlrim(y=y_cbind,
                                 mu=~ stable_cdf2(a+b*dose+rand, c(alpha, 0, 1/sqrt(2), 0)),
                                 pmu=c(a=0,b=0, alpha=2),
                                 pmix=c(alpha=2, scl=1),
                                 p_uppb = c(Inf ,  Inf, 2, Inf),
                                 p_lowb = c(-Inf, -Inf, 2,   0+1e-5),
                                 distribution="binomial",
                                 nest=id,
                                 random="rand",
                                 mixture="libstableR-subgauss-scl",
                                 ooo=TRUE,
                                 compute_hessian = FALSE,
                                 compute_kkt = FALSE
)

fit_PPN_libstableR_scl

```

```
> fit_PPN_libstableR_scl
                a         b alpha      scl    value fevals gevals niter
nlminb -0.4832709 0.3560649     2 1.526264 8.966632     15     52    13
       convcode kkt1 kkt2 xtimes
nlminb        0   NA   NA   5.45
```

**Comment by swihart** [comment]
From the cluster...still getting convcode 1 for alpha==2.01 after making iterlim=5000.

```
> sAll
  alpha         a0        a1     gam_h      alp     gam_g    value fevals
1  1.00 -0.2736897 0.6862525 0.7071068 1.000000  3.291157 309.3916     15
2  2.00 -0.3743794 0.6592300 0.7071068 2.000000  1.358630 316.0192     18
3  2.01 -1.9997529 2.9994805 0.7071068 1.001000  2.999715 329.0000    130
4  2.02 -1.6807357 4.1362839 3.8469433 1.159966 14.430400 308.9461     75
  gevals niter convcode kkt1 kkt2   xtimes      time time_num time_units
1     48    12        0   NA   NA   22.008 31.9 secs     31.9       secs
2     53    13        0   NA   NA   28.073 36.8 secs     36.8       secs
3    152    29        1   NA   NA 3571.203 33.8 mins     33.8       mins
4    369    48        0   NA   NA 5535.988 48.2 mins     48.2       mins
```


So, I'm guessing, that optimx is giving convcode 1 if the **evaluations** of the objective function exceeds the  `eval.max` argument in nlminb?  
```
eval.max
Maximum number of evaluations of the objective function allowed. Defaults to 200.
```

HOLD ON HOLD ON.  optimx says you get a convcode 1 if `maxit` (not `itnmax`) is exceeded.

```
convcode An integer code. 0 indicates successful convergence. Various methods may or may not
return sufficient information to allow all the codes to be specified. An incomplete list of codes
includes
1 indicates that the iteration limit maxit had been reached.
```

where:

```
maxit The maximum number of iterations. Defaults to 100 for the derivative-based methods, and
500 for "Nelder-Mead".
```

So, I suppose I need to make iterlim connect to maxit in the control() of optimx as well???  Try it.

**Comment by swihart** [comment]
also remember:  I think these troubles started after doing cdf2/pdf2 business.... so that's something to try if maxit adjustment doesn't fix it.  Also -- starting values.  Quickest thing to  do is change starting values.  Just did.  Just resubmitted.  Let's see.

Changed starting values and ....

```
> sAll
  alpha         a0        a1     gam_h      alp     gam_g    value fevals
1  1.00 -0.2736897 0.6862525 0.7071068 1.000000  3.291157 309.3916     15
2  2.00 -0.3743794 0.6592300 0.7071068 2.000000  1.358630 316.0192     18
3  2.01 -1.9997529 2.9994805 0.7071068 1.001000  2.999715 329.0000    101
4  2.02 -1.6807357 4.1362839 3.8469433 1.159966 14.430400 308.9461     75
  gevals niter convcode kkt1 kkt2   xtimes      time time_num time_units
1     48    12        0   NA   NA   22.122 31.5 secs     31.5       secs
2     53    13        0   NA   NA   28.126 36.4 secs     36.4       secs
3    117    22        1   NA   NA 5496.598   49 mins       49       mins
4    369    48        0   NA   NA 5800.130 50.8 mins     50.8       mins

```

Nothing changed!  This is because I failed to change `pmix`, only changed `pmu`.  Must remember this!  Okay, change pmix to alpha=1.2 and gam_g=5 and got the following for the 
"alpha=2.01" scenario in an interactive session, trace = 1 so I can see message that is printed.

<img width="711" alt="screen shot 2017-11-09 at 4 36 00 am" src="https://user-images.githubusercontent.com/4581196/32603461-d50ba18a-c517-11e7-8ad6-0bcd6ef337ab.png">


Good news:  convergence!  Reasonable estimates!
Bad news:  need to rerun interactive session with old values to see why I was getting a 1 for convcode.
One thought is that when freely estimating alpha, don't have it start on alpha=1.  Or, maybe this is just "one of those random things"....

**Comment by swihart** [comment]
Okay, so it turns out it was not an iterlim / itermax thing at all!  It was false convergence!

```
> SSS_PPN_pcauchy <- gnlrim(   y = y2,
+                                   distribution = "binomial",
+                                   mixture = "libstableR-subgauss-scl",
+                                   random = "rand",
+                                   nest = id,
+                                   mu = ~ stable_cdf2(
+ ## marginal coefficients
+ (a0 + a1*group)*
+ ## phi^-1
+ (gam_h/(gam_h^alp+ gam_g^alp)^(1/alp)  )^(-1) +
+ ## alpha-stable distributed random intercept
+ rand,
+                                                              c(alp,0,gam_h,0),0),
+ 
+ 
+                    p_uppb = c(a0= 99.0, a1= 50, gam_h=1/sqrt(2), alp=2-1e-2 , gam_g=Inf    ),
+                       pmu = c(a0=- 2.0, a1=  3, gam_h=1/sqrt(2), alp=1      , gam_g= 3     ),
+                    p_lowb = c(a0=-99.0, a1=-50, gam_h=1/sqrt(2), alp=0+1e-1, gam_g= 0+1e-1),
+                      pmix = c(alp=1, gam_g=3),
+ 
+                       ooo = TRUE,
+           compute_hessian = FALSE,
+               compute_kkt = FALSE,
+ 
+                   iterlim = 5000,
+ trace = 1
+                           )
fn is  fn 
Looking for method =  nlminb 
Function has  5  arguments
par[ 1 ]:  -99   <? -2   <? 99     In Bounds      
par[ 2 ]:  -50   <? 3   <? 50     In Bounds   In Bounds     
par[ 3 ]:  0.7071068   <? 0.7071068   <? 0.7071068     In Bounds   In Bounds   In Bounds    
par[ 4 ]:  0.1   <? 1   <? 1.99     In Bounds   In Bounds   In Bounds   In Bounds   
par[ 5 ]:  0.1   <? 3   <? Inf     In Bounds   In Bounds   In Bounds   In Bounds   In Bounds  
Analytic gradient not made available.
Analytic Hessian not made available.
Scale check -- log parameter ratio= 0.6276363   log bounds ratio= 2.020203 
Method:  nlminb 
print(subprobit.df)
  0:     329.02512: -2.00000  3.00000 0.707107  1.00000  3.00000
  1:     329.00257: -1.99978  2.99953 0.707107  1.00090  2.99974
  2:     329.00142: -1.99977  2.99951 0.707107  1.00094  2.99973
  3:     329.00119: -1.99977  2.99951 0.707107  1.00095  2.99973
  4:     329.00073: -1.99976  2.99950 0.707107  1.00097  2.99972
  5:     329.00055: -1.99976  2.99949 0.707107  1.00098  2.99972
  6:     329.00018: -1.99975  2.99948 0.707107  1.00099  2.99972
  7:     329.00003: -1.99975  2.99948 0.707107  1.00100  2.99972
  8:     329.00000: -1.99975  2.99948 0.707107  1.00100  2.99972
  9:     329.00000: -1.99975  2.99948 0.707107  1.00100  2.99972
 10:     329.00000: -1.99975  2.99948 0.707107  1.00100  2.99972
 11:     328.99999: -1.99975  2.99948 0.707107  1.00100  2.99972
 12:     328.99999: -1.99975  2.99948 0.707107  1.00100  2.99972
 13:     328.99999: -1.99975  2.99948 0.707107  1.00100  2.99972
 14:     328.99999: -1.99975  2.99948 0.707107  1.00100  2.99972
 15:     328.99999: -1.99975  2.99948 0.707107  1.00100  2.99972
 16:     328.99999: -1.99975  2.99948 0.707107  1.00100  2.99972
 17:     328.99999: -1.99975  2.99948 0.707107  1.00100  2.99972
 18:     328.99999: -1.99975  2.99948 0.707107  1.00100  2.99972
 19:     328.99999: -1.99975  2.99948 0.707107  1.00100  2.99972
 20:     328.99999: -1.99975  2.99948 0.707107  1.00100  2.99972
 21:     328.99999: -1.99975  2.99948 0.707107  1.00100  2.99972
 22:     328.99999: -1.99975  2.99948 0.707107  1.00100  2.99972
Post processing for method  nlminb 
Save results from method  nlminb 
$par
        a0         a1      gam_h        alp      gam_g 
-1.9997529  2.9994805  0.7071068  1.0010000  2.9997151 

$message
[1] "false convergence (8)"

$convcode
[1] 1

$value
[1] 329

$fevals
function 
     101 

$gevals
gradient 
     117 

$nitns
[1] 22

$kkt1
[1] NA

$kkt2
[1] NA

$xtimes
user.self 
 6883.671 

Assemble the answers

```

**Comment by swihart** [comment]
Hey, just rerun real-quick with method = nmkb....

Just did.  Can't do it because the way I hold parameters constant means initial parms are on the bounds.

try method = bobyqa -- "bobyqa failed for current problem"

try method = hjkb -- finished in seconds with a successful convergence code, none of parms are updated.  YIKES.

**Comment by swihart** [comment]
I think what I learned from this.... is...

**New issue:  how to handle masks better**
  * might want to try one of those masks procedures (see new issue about nlminb default)
 


**Rework flow**
  *  swarm_file in order of alpha -- in fact do a fake out for first job so that the `_1` suffix listed via squeue matches the alpha_in (have email confirmation that I can't specify job suffixes, which means I can't get clever with my ordering in straightfoward way)
  *  put explicit print() statement around the function call so that any error messages/trace info will print to the .Rout files 
  *  Don't compute hessian for the 202 alphas
  *  swarm -t=16 seems to speed things up
  *  Isolate the convcode 1 after rbinds
  *  Look at message codes in .Rout files (which are named with corresponding alpha) and/or just resubmit those with convcode 1 with different starting values
  *  Then adjust bounds and calculate hessian/start checks etc for the 2.01 / 2.02 scenario.

**Comment by swihart** [comment]
I am about to log on to biowulf and combine the results from the `nohess` approach with starting values that didn't have alpha=1:

This is interesting.  Remember this is for the smaller dataset.  The default walltime was 2 hours.  I have 149 out of 202 possible RDS.  2 of them are the specialty, free-estimating alpha, so let's adjust and say 147 out of 200.  Nothing below alpha=.40 finished.  but this might be enough info to make the graphs and see a minimum.   Let's check convcodes and actual run times:

  *  Only 6 of the 147 went over an hour.  The other 141 was under an hour.
  *  16% (or 24 of the 147) had a convcode of 1.  The rest had 0.

What about our alpha==2.01 scenario?

  *  22.3 minutes.
  *  convcode 0.

Dang!!!!!!!!!!!!!!!!!!

**Comment by swihart** [comment]
What's interesting in graph below is that a bunch of `convcode==1` fall in-line.  The 2nd graph below it just shows them removed.
![dt01-1](https://user-images.githubusercontent.com/4581196/32729086-9ebc2370-c850-11e7-84c6-cd89a6d2783f.png)
![dt01-2](https://user-images.githubusercontent.com/4581196/32729087-9ec8b982-c850-11e7-822d-2a1dcbe895ec.png)

**Comment by swihart** [comment]
Let's try this approach for the big dataset.  On the cluster.

... okay ... up and running.  Alpha 1 and 2 are in less than 2 minutes, convcodes 0, maxlik looking like they did in the `nlm` anlaysis... looks like (from the reports I made in the past) that the large dataset, with nlm and hessian etc, the first results didn't come in for 4 hours...we'll see if this nohess nlminb approach can be faster...

About an hour in and nothing aside from CCC and PPN are in.  I'm doing peeking at the heads of some .Rout files and I'm really shocked at how far off the likelihood is from the initial estimates which made me think that maybe we can smarten up the work flow even more.  Since Cauchy and Probit run so fast, maybe the work flow should start by running those two, and then possible adjust all the other alphas`pmu` and `pmix` starting values based on that.  IT will not be as simple as an interpolation because the coefficients don't vary linearly with alpha, but certainly going with the final values from a Cauchy run or probit run at least for the nearby alphas might be helpful...

So, new proposed work flow:

**Rework flow**
  *  swarm_file in smart order of alpha go ahead and launch alpha = 1, 2, 2.01, 2.02 scenarios.
  *  Once alpha=1,2 come back, look at values.  Consider adjusting starting values based on these before launching 198 more alphas.
  * be sure .Rout filename corresponds to alpha
  *  put explicit print() statement around the function call so that any error messages/trace info will print to the .Rout files 
  *  Don't compute hessian for the 202 alphas
  *  swarm -t=16 seems to speed things up
  *  Isolate the convcode 1 after rbinds
  *  Look at message codes in .Rout files (which are named with corresponding alpha) and/or just resubmit those with convcode 1 with different starting values
  *  Then adjust bounds and calculate hessian/start checks etc for the 2.01 / 2.02 scenario.  Although, you did submit it first, blindly and maybe have it running this whole time, which isn't bad if the resources are available... but if resources short, don't submit 2.01/2.02 scenario until you make the graphs for maxlik and parameter values.

**Comment by swihart** [comment]
Sweet.  Around 1 hour mark some start trickling in (like alpha 1.49, 1.50)...

conv codes are 0.

Around 1.5 hour mark 65 .RDS files marching in.  Sweet.  hopefully we'll have a gush before the 2 hour walltime is up.

As of yet, none are alpha < 1.00 though...

I might extend wall time, because, my graph is not going to have a minimum if there aren't any alphas below 1. (!)

I extended wall time...right now we are 10 min from 2 hour running time and we have 81 .RDS files but none below alpha = 1 (!)

Okay, it is official.  At 2 hour mark we have 88 RDS files, none from alpha < 1.  Good thing I extended time.

**Comment by swihart** [comment]
For a minute .. wanted to look at .Routs of the **small** **nohess** approach and see if it is iteration limits or convergence issues.

Emailed Ravi.  Looks like all 24 came back quickly (so not iteration limit issue) and 23 had (8) and 1 had (9) message:

> 
Just a little more info:  I just ran optimx(method="nlminb", trace=1) for about 150 runs of a profile likelihood.  I received 24 of those returning `$convcode 1`.  Of those 24 the `$message` from nlminb (which I can see with trace=1) for 23 of them is "false convergence (8)" and for 1 of them is "function evaluation limit reached without convergence (9)".  All 24 returned quickly relative to the others that successfully converged and did not approach the 50000 iteration limit I imposed.
